plutus-contract/src/Language/Plutus/Contract/StateMachine.hs

```diff
@@ -0,0 +1,206 @@
+        , vaOtherPayments :: UnbalancedTx
+        -- ^ Any other payments
+        }
+
+instance Semigroup ValueAllocation where
+    l <> r =
+        ValueAllocation
+            { vaOwnAddress = vaOwnAddress l <> vaOwnAddress r
+            , vaOtherPayments = vaOtherPayments l <> vaOtherPayments r
+            }
+
+instance Monoid ValueAllocation where
+    mappend = (<>)
+    mempty  = ValueAllocation mempty mempty
+
+-- | Client-side definition of a state machine.
+data StateMachineClient s i = StateMachineClient
+    { scInstance :: SM.StateMachineInstance s i
+    -- ^ The instance of the state machine, defining the machine's transitions,
+    --   its final state and its check function.
+    , scPayments :: s -> i -> Value -> ValueAllocation
+    -- ^ A function that determines the 'ValueAllocation' of each transition,
+    --   given the value currently locked by the contract.
+    , scChooser  :: [SM.OnChainState s i] -> Either (SMContractError s i) (SM.OnChainState s i)
+    -- ^ A function that chooses the relevant on-chain state, given a list of
+    --   all potential on-chain states found at the contract address.
+    }
+
+-- | A state chooser function that fails if confronted with anything other
+--   than exactly one output
+defaultChooser ::
+    forall state input
+    . [SM.OnChainState state input]
+    -> Either (SMContractError state input) (SM.OnChainState state input)
+defaultChooser [x] = Right x
+defaultChooser xs  =
+    let msg = "Found " <> show (length xs) <> " outputs, expected 1"
+    in Left (ChooserError (Text.pack msg))
+
+-- | A state machine client with the 'defaultChooser' function
+mkStateMachineClient ::
+    forall state input
+    . SM.StateMachineInstance state input
+    -> (state -> input -> Value -> ValueAllocation)
+    -> StateMachineClient state input
+mkStateMachineClient inst payments =
+    StateMachineClient
+        { scInstance = inst
+        , scPayments = payments
+        , scChooser  = defaultChooser
+        }
+
+getOnChainState :: (AsSMContractError e state i, PlutusTx.IsData state, HasUtxoAt schema) => StateMachineClient state i -> Contract schema e (SM.OnChainState state i)
+getOnChainState StateMachineClient{scInstance, scChooser} = do
+    utxo <- utxoAt (SM.machineAddress scInstance)
+    let states = SM.getStates scInstance utxo
+    either (throwing _SMContractError) pure $ scChooser states
+
+-- | Run one step of a state machine, returning the new state.
+runStep ::
+    forall e state schema input.
+    ( AsSMContractError e state input
+    , AsContractError e
+    , PlutusTx.IsData state
+    , PlutusTx.IsData input
+    , HasUtxoAt schema
+    , HasWriteTx schema
+    , HasTxConfirmation schema
+    )
+    => StateMachineClient state input
+    -- ^ The state machine
+    -> input
+    -- ^ The input to apply to the state machine
+    -> Contract schema e state
+runStep smc input = do
+    (typedTx, newState, ValueAllocation{vaOtherPayments}) <- mkStep smc input
```

Comment:
I was initially confused about why we don't need to handle the payments to our own address here - worth a comment.